scipy/optimize/zeros.py

```diff
@@ -163,54 +163,50 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
         # Newton-Rapheson method
         # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
         # so it still works if x0 is complex.
-        p0 = 1.0 * x0
-        fder2 = 0
+        p0 = 1.0 * asarray(x0)  # convert to ndarray
         for iter in range(maxiter):
             myargs = (p0,) + args
-            fder = fprime(*myargs)
-            if fder == 0:
+            fder = asarray(fprime(*myargs))  # convert to ndarray
+            if (fder == 0).any():
                 msg = "derivative was zero."
                 warnings.warn(msg, RuntimeWarning)
                 return p0
-            fval = func(*myargs)
-            if fprime2 is not None:
-                fder2 = fprime2(*myargs)
-            if fder2 == 0:
+            fval = asarray(func(*myargs))  # convert to ndarray
+            if fprime2 is None:
                 # Newton step
                 p = p0 - fval / fder
             else:
-                # Parabolic Halley's method
-                discr = fder ** 2 - 2 * fval * fder2
-                if discr < 0:
-                    p = p0 - fder / fder2
-                else:
-                    p = p0 - 2*fval / (fder + sign(fder) * sqrt(discr))
-            if abs(p - p0) < tol:
+                fder2 = asarray(fprime2(*myargs))  # convert to ndarray
+                # Halley's method
+                # https://en.wikipedia.org/wiki/Halley%27s_method
+                p = p0 - 2 * fval * fder / (2 * fder ** 2 - fval * fder2)
+            if np_abs(p - p0).max() < tol:
                 return p
             p0 = p
     else:
         # Secant method
-        p0 = x0
-        if x0 >= 0:
-            p1 = x0*(1 + 1e-4) + 1e-4
-        else:
-            p1 = x0*(1 + 1e-4) - 1e-4
-        q0 = func(*((p0,) + args))
-        q1 = func(*((p1,) + args))
+        p0 = asarray(x0)
+        dx = finfo(float).eps**0.33
+        dp = where(p0 >= 0, dx, -dx)
+        p1 = p0 * (1 + dx) + dp
+        q0 = asarray(func(*((p0,) + args)))
+        q1 = asarray(func(*((p1,) + args)))
         for iter in range(maxiter):
-            if q1 == q0:
-                if p1 != p0:
-                    msg = "Tolerance of %s reached" % (p1 - p0)
+            divide_by_zero = (q1 == q0)
+            if divide_by_zero.any():
```

Comment:
here as well, wouldn't it be better to continue for all other elements?

Reply:
See my previous answer in the `newton` branch. My ethic was to make the smallest changes possible to add some new features. If users demand more features, then I think we should discuss the trade offs.

Reply:
The halting conditions are the biggest thing to think about here. This way is simple, but I'm sure that *someone* is going to yell at us in the future about evaluating their callback function too many times. Might be worth seeing how messy the indexing would look.